### PR TITLE
android: Database encryption support with a passphrase

### PR DESCRIPTION
**Pull request by avently**
To test add this into `MainActivity` `onCreate()`
```kotlin
Cryptor.test(usePreferences = true)
```

### CODE REVIEW

File: apps/android/app/src/main/java/chat/simplex/app/SimplexApp.kt [hunk 1]
```diff
@@ -35,9 +36,18 @@ external fun chatRecvMsgWait(ctrl: ChatCtrl, timeout: Int): String
...
 class SimplexApp: Application(), LifecycleEventObserver {
-  val chatController: ChatController by lazy {
-    val ctrl = chatInit(getFilesDirectory(applicationContext))
-    ChatController(ctrl, ntfManager, applicationContext, appPreferences)
+  lateinit var chatController: ChatController
+
+  fun initChatController(useKey: String? = null) {
+    val dbKey = useKey ?: DatabaseUtils.getDatabaseKey() ?: ""
+    val res = DatabaseUtils.migrateChatDatabase(dbKey)
+    val ctrl = chatInitKey(getFilesDirectory(applicationContext), dbKey)
```
Comment (epoberezkin) [inline apps/android/app/src/main/java/chat/simplex/app/SimplexApp.kt:44]:
```suggestion
    val ctrl = chatInitKey(getFilesDirectory(applicationContext), "") // dbKey)
```

for now we can just pass an empty key

File: apps/android/app/src/main/java/chat/simplex/app/model/SimpleXAPI.kt [hunk 2]
```diff
@@ -106,6 +106,11 @@ class AppPreferences(val context: Context) {
   val networkTCPKeepCnt = mkIntPreference(SHARED_PREFS_NETWORK_TCP_KEEP_CNT, KeepAliveOpts.defaults.keepCnt)
   val incognito = mkBoolPreference(SHARED_PREFS_INCOGNITO, false)
 
+  val storeDBPassphrase = mkBoolPreference(SHARED_PREFS_STORE_DB_PASSPHRASE, true)
+  val initialRandomDBPassphrase = mkBoolPreference(SHARED_PREFS_INITIAL_RANDOM_DB_PASSPHRASE, false)
+  val encryptedDBPassphrase = mkStrPreference(SHARED_PREFS_ENCRYPTED_DB_PASSPHRASE, null)
+  val initializationVectorDBPassphrase = mkStrPreference(SHARED_PREFS_INITIALIZATION_VECTOR_DB_PASSPHRASE, null)
```
Comment (epoberezkin) [inline apps/android/app/src/main/java/chat/simplex/app/model/SimpleXAPI.kt:112]:
```suggestion
  val encryptedDBPassphrase = mkStrPreference(SHARED_PREFS_DB_PASSPHRASE_ENCRYPTED, null)
  val initializationVectorDBPassphrase = mkStrPreference(SHARED_PREFS_DB_PASSPHRASE_IV, null)
```

File: apps/android/app/src/main/res/values-ru/strings.xml [hunk 3]
```diff
@@ -67,12 +67,21 @@
     <string name="periodic_notifications">Периодические уведомления</string>
     <string name="periodic_notifications_disabled">Периодические уведомления выключены!</string>
     <string name="periodic_notifications_desc">Приложение периодически получает новые сообщения — это потребляет несколько процентов батареи в день. Приложение не использует push уведомления — данные не отправляются с вашего устройства на сервер.</string>
+    <string name="enter_passphrase_notification_title">Введите пароль</string>
+    <string name="enter_passphrase_notification_desc">Для получения уведомлений, пожалуйста, введите пароль от базы данных</string>
+    <string name="database_initialization_error_title">Can\'t initialize the database</string>
+    <string name="database_initialization_error_desc">The database is not working correctly. Tap to learn more</string>
```
Comment (jr-simplex) [inline apps/android/app/src/main/res/values-ru/strings.xml:73]:
```suggestion
    <string name="database_initialization_error_title">Ошибка при инициализации базы данных</string>
    <string name="database_initialization_error_desc">Некорректная работа базы данных. Нажмите чтобы узнать больше</string>
```

File: apps/android/app/src/main/res/values-ru/strings.xml [hunk 4]
```diff
@@ -291,7 +302,7 @@
     <!-- settings - SettingsView.kt -->
     <string name="your_settings">Настройки</string>
     <string name="your_simplex_contact_address">Ваш <xliff:g id="appName">SimpleX</xliff:g> адрес</string>
-    <string name="database_export_and_import">Экспорт и импорт архива чата</string>
+    <string name="database_passphrase_and_export">Защита данных чата</string>
```
Comment (jr-simplex) [inline apps/android/app/src/main/res/values-ru/strings.xml:305]: not sure about it
```suggestion
    <string name="database_passphrase_and_export">Управление базой данных</string>
```

File: apps/android/app/src/main/res/values-ru/strings.xml [hunk 5]
```diff
@@ -517,6 +528,7 @@
     <string name="chat_is_running">Чат запущен</string>
     <string name="chat_is_stopped">Чат остановлен</string>
     <string name="chat_database_section">АРХИВ ЧАТА</string>
+    <string name="database_passphrase">Пароль данных чата</string>
```
Comment (jr-simplex) [inline apps/android/app/src/main/res/values-ru/strings.xml:531]:
```suggestion
    <string name="database_passphrase">Пароль базы данных</string>
```

File: apps/android/app/src/main/res/values-ru/strings.xml [hunk 6]
```diff
@@ -526,6 +538,8 @@
     <string name="stop_chat_question">Остановить чат?</string>
     <string name="stop_chat_to_export_import_or_delete_chat_database">Остановите чат, чтобы экспортировать или импортировать архив чата или удалить данные чата. Вы не сможете получать и отправлять сообщения, пока чат остановлен.</string>
     <string name="stop_chat_confirmation">Остановить</string>
+    <string name="set_password_to_export">Установите пароль</string>
+    <string name="set_password_to_export_desc">Данные чата зашифрованы случайным паролем. Пожалуйста, поменяйте его перед экспортом.</string>
```
Comment (jr-simplex) [inline apps/android/app/src/main/res/values-ru/strings.xml:542]:
```suggestion
    <string name="set_password_to_export_desc">База данных зашифрована случайным паролем. Пожалуйста, поменяйте его перед экспортом.</string>
```

File: apps/android/app/src/main/res/values-ru/strings.xml [hunk 7]
```diff
@@ -541,7 +555,53 @@
     <string name="restart_the_app_to_create_a_new_chat_profile">Перезапустите приложение, чтобы создать новый профиль.</string>
     <string name="you_must_use_the_most_recent_version_of_database">Используйте самую последнюю версию архива чата и ТОЛЬКО на одном устройстве, иначе вы можете перестать получать сообщения от некоторых контактов.</string>
     <string name="stop_chat_to_enable_database_actions">Остановите чат, чтобы разблокировать операции с архивом чата.</string>
-    <string name="restart_the_app_to_use_new_chat_database">Перезапустите приложение, чтобы использовать новый архив чата.</string>
+
+    <!-- DatabaseEncryptionView.kt -->
+    <string name="save_passphrase_in_keychain">Сохраните пароль в Keystore</string>
+    <string name="database_encrypted">Данные чата зашифрованы!</string>
+    <string name="error_encrypting_database">Ошибка при шифровании данных</string>
```
Comment (jr-simplex) [inline apps/android/app/src/main/res/values-ru/strings.xml:562]:
```suggestion
    <string name="save_passphrase_in_keychain">Сохранить пароль в Keystore</string>
    <string name="database_encrypted">База данных зашифрована!</string>
    <string name="error_encrypting_database">Ошибка при шифровании базы данных</string>
```

File: apps/android/app/src/main/res/values-ru/strings.xml [hunk 8]
```diff
@@ -541,7 +555,53 @@
...
+    <string name="remove_passphrase_from_keychain">Удалить пароль из Keystore?</string>
+    <string name="notifications_will_be_hidden">Уведомления будут работать только до остановки приложения!</string>
```
Comment (jr-simplex) [inline apps/android/app/src/main/res/values-ru/strings.xml:564]:
```suggestion
    <string name="notifications_will_be_hidden">Уведомления будут работать только пока приложение открыто!</string>
```

File: apps/android/app/src/main/res/values-ru/strings.xml [hunk 9]
```diff
@@ -541,7 +555,53 @@
...
+    <string name="remove_passphrase">Удалить</string>
+    <string name="encrypt_database">Зашифровать</string>
+    <string name="update_database">Поменять</string>
+    <string name="current_passphrase">Текущий пароль…</string>
+    <string name="new_passphrase">Новый пароль…</string>
+    <string name="confirm_new_passphrase">Подтвердить новый пароль…</string>
```
Comment (jr-simplex) [inline apps/android/app/src/main/res/values-ru/strings.xml:570]:
```suggestion
    <string name="confirm_new_passphrase">Подтвердите новый пароль…</string>
```

File: apps/android/app/src/main/res/values-ru/strings.xml [hunk 10]
```diff
@@ -541,7 +555,53 @@
...
+    <string name="update_database_passphrase">Поменять пароль данных</string>
```
Comment (jr-simplex) [inline apps/android/app/src/main/res/values-ru/strings.xml:571]:
```suggestion
    <string name="update_database_passphrase">Поменять пароль базы данных</string>
```

File: apps/android/app/src/main/res/values-ru/strings.xml [hunk 11]
```diff
@@ -541,7 +555,53 @@
...
+    <string name="enter_correct_current_passphrase">Пожалуйста, введите правильный пароль.</string>
+    <string name="database_is_not_encrypted">Данные чата не зашифрованы - установите пароль чтобы защитить данные.</string>
```
Comment (jr-simplex) [inline apps/android/app/src/main/res/values-ru/strings.xml:573]:
```suggestion
    <string name="database_is_not_encrypted">База данных чата не зашифрована - установите пароль, чтобы защитить ваши данные.</string>
```

File: apps/android/app/src/main/res/values-ru/strings.xml (hunk 8 above)
Comment (epoberezkin) [reply apps/android/app/src/main/res/values-ru/strings.xml:564]: this is not the same

File: apps/android/app/src/main/res/values-ru/strings.xml [hunk 12]
```diff
@@ -541,7 +555,53 @@
...
+    <string name="keychain_is_storing_securely">Android Keystore используется для безопасного хранения пароля - это позволяет получать нотификации из фонового сервиса.</string>
```
Comment (jr-simplex) [inline apps/android/app/src/main/res/values-ru/strings.xml]: I think "notifications" were translated as "уведомления" somewhere else in the translations.. Both are okay but I'd stick with one
```suggestion
    <string name="keychain_is_storing_securely">Android Keystore используется для безопасного хранения пароля - это позволяет получать уведомления в фоновом режиме.</string>
```

File: apps/android/app/src/main/res/values-ru/strings.xml [hunk 13]
```diff
@@ -541,7 +555,53 @@
...
+    <string name="encrypted_with_random_passphrase">Данные чата зашифрованы случайным паролем, вы можете его поменять.</string>
```
Comment (jr-simplex) [inline apps/android/app/src/main/res/values-ru/strings.xml:575]:
```suggestion
    <string name="encrypted_with_random_passphrase">База данных зашифрована случайным паролем, вы можете его поменять.</string>
```

File: apps/android/app/src/main/res/values-ru/strings.xml [hunk 14]
```diff
@@ -541,7 +555,53 @@
...
+    <string name="impossible_to_recover_passphrase"><b>Внимание</b>: вы не сможете восстановить или поменять пароль, если вы его потеряете.</string>
+    <string name="keychain_allows_to_receive_ntfs">Android Keystore will be used to securely store passphrase after you restart the app or change passphrase - it will allow receiving notifications.</string>
```
Comment (jr-simplex) [inline apps/android/app/src/main/res/values-ru/strings.xml]:
```suggestion
    <string name="keychain_allows_to_receive_ntfs">Пароль будет безопасно сохранен в Android Keystore после перезапуска приложения или изменения пароля — это позволит получать уведомления.</string>
```

File: apps/android/app/src/main/res/values-ru/strings.xml [hunk 15]
```diff
@@ -541,7 +555,53 @@
...
+    <string name="you_have_to_enter_passphrase_every_time">Вы должны вводить пароль каждый раз при запуске чата - пароль не сохранен на устройстве.</string>
```
Comment (jr-simplex) [inline apps/android/app/src/main/res/values-ru/strings.xml:578]:
```suggestion
    <string name="you_have_to_enter_passphrase_every_time">Пароль не сохранен на устройстве — вы будете должны ввести его при каждом запуске чата.</string>
```

File: apps/android/app/src/main/res/values-ru/strings.xml [hunk 16]
```diff
@@ -541,7 +555,53 @@
...
+    <string name="database_will_be_encrypted">Данные чата будут зашифрованы.</string>
+    <string name="database_will_be_encrypted_and_passphrase_stored">Данные чата будут зашифрованы и пароль сохранен в Keystore.</string>
+    <string name="database_encryption_will_be_updated">Пароль данных чата будет изменен и сохранен в Keystore.</string>
+    <string name="database_passphrase_will_be_updated">Пароль данных чата будет изменен.</string>
+    <string name="store_passphrase_securely">Пожалуйста, надежно сохраните пароль, вы НЕ сможете его поменять, если потеряете.</string>
+    <string name="store_passphrase_securely_without_recover">Пожалуйста, надежно сохраните пароль, вы НЕ сможете открыть чат, если вы потеряете пароль.</string>
+
+    <!-- DatabaseErrorView.kt -->
+    <string name="wrong_passphrase">Неправильный пароль данных</string>
+    <string name="encrypted_database">Зашифрованные данные чата</string>
```
Comment (jr-simplex) [inline apps/android/app/src/main/res/values-ru/strings.xml:590]:
```suggestion
    <string name="encrypted_database">Данные зашифрованы</string>
```

File: apps/android/app/src/main/res/values-ru/strings.xml [hunk 17]
```diff
@@ -541,7 +555,53 @@
...
+    <string name="change_database_passphrase_question">Поменять пароль данных?</string>
+    <string name="database_will_be_encrypted">Данные чата будут зашифрованы.</string>
+    <string name="database_will_be_encrypted_and_passphrase_stored">Данные чата будут зашифрованы и пароль сохранен в Keystore.</string>
+    <string name="database_encryption_will_be_updated">Пароль данных чата будет изменен и сохранен в Keystore.</string>
+    <string name="database_passphrase_will_be_updated">Пароль данных чата будет изменен.</string>
+    <string name="store_passphrase_securely">Пожалуйста, надежно сохраните пароль, вы НЕ сможете его поменять, если потеряете.</string>
+    <string name="store_passphrase_securely_without_recover">Пожалуйста, надежно сохраните пароль, вы НЕ сможете открыть чат, если вы потеряете пароль.</string>
+
+    <!-- DatabaseErrorView.kt -->
+    <string name="wrong_passphrase">Неправильный пароль данных</string>
```
Comment (jr-simplex) [inline apps/android/app/src/main/res/values-ru/strings.xml:589]:
```suggestion
    <string name="wrong_passphrase">Неправильный пароль</string>
```

File: apps/android/app/src/main/res/values-ru/strings.xml [hunk 18]
```diff
@@ -541,7 +555,53 @@
...
+    <string name="encrypted_database">Зашифрованные данные чата</string>
+    <string name="database_error">Ошибка данных</string>
+    <string name="keychain_error">Ошибка Keystore</string>
+    <string name="passphrase_is_different">Пароль данных отличается от пароля сохрененного в Keystore.</string>
```
Comment (jr-simplex) [inline apps/android/app/src/main/res/values-ru/strings.xml:593]:
```suggestion
    <string name="passphrase_is_different">Пароль отличается от сохраненного в Keystore.</string>
```

File: apps/android/app/src/main/res/values-ru/strings.xml [hunk 19]
```diff
@@ -541,7 +555,53 @@
...
+    <string name="file_with_path">Файл: %s</string>
+    <string name="database_passphrase_is_required">Введите пароль данных чтобы открыть чат.</string>
```
Comment (jr-simplex) [inline apps/android/app/src/main/res/values-ru/strings.xml:595]:
```suggestion
    <string name="database_passphrase_is_required">Введите пароль чтобы открыть чат.</string>
```

**Pull request merged**